File: db4on/reference/db4o-reference-chapters/src/com/db4odoc/f1/identity/IdentityExample.cs
/* Copyright (C) 2004 - 2006 db4objects Inc. http://www.db4o.com */
using System;
using System.IO;
using com.db4o;
using com.db4o.query;

namespace com.db4odoc.f1.identity
{
	public class IdentityExample
	{
		public readonly static string YapFileName = "formula1.yap";

		public static void Main(String[] args) 
		{
			CheckUniqueness();
			CheckReferenceCache();
			CheckReferenceCacheWithPurge();
			TestBind();
		
			TestCopyingWithPurge();
		}
		// end Main

		public static void SetObjects()
		{
			File.Delete(YapFileName);
			ObjectContainer db = Db4o.OpenFile(YapFileName);
			try 
			{
				Car car = new Car("BMW", new Pilot("Rubens Barrichello"));
				db.Set(car);
				car = new Car("Ferrari", new Pilot("Michael Schumacher"));
				db.Set(car);
			} 
			finally 
			{
				db.Close();
			}
		}
		// end SetObjects
	
		public static void CheckUniqueness()
		{
			SetObjects();
			ObjectContainer db = Db4o.OpenFile(YapFileName);
			try 
			{
				ObjectSet cars = db.Get(typeof(Car));
				Car car = (Car)cars[0];
				String pilotName = car.Pilot.Name;
				ObjectSet pilots = db.Get(new Pilot(pilotName));
				Pilot pilot = (Pilot)pilots[0];
				System.Console.WriteLine("Retrieved objects are identical: " + (pilot == car.Pilot));
			} 
			finally 
			{
				db.Close();
			}
		}
		// end CheckUniqueness
	
		public static void CheckReferenceCache()
		{
			SetObjects();
			ObjectContainer db = Db4o.OpenFile(YapFileName);
			try 
			{
				ObjectSet pilots = db.Get(typeof(Pilot));
				Pilot pilot = (Pilot)pilots[0];
				String pilotName = pilot.Name;
				pilot.Name = "new name";
				System.Console.WriteLine("Retrieving pilot by name: " + pilotName);
				ObjectSet pilots1 = db.Get(new Pilot(pilotName));
				ListResult(pilots1);
			} 
			finally 
			{
				db.Close();
			}
		}
		// end CheckReferenceCache
	
		public static void CheckReferenceCacheWithPurge()
		{
			SetObjects();
			ObjectContainer db = Db4o.OpenFile(YapFileName);
			try 
			{
				ObjectSet pilots = db.Get(typeof(Pilot));
				Pilot pilot = (Pilot)pilots[0];
				String pilotName = pilot.Name;
				pilot.Name = "new name";
				System.Console.WriteLine("Retrieving pilot by name: " + pilotName);
				long pilotID = db.Ext().GetID(pilot);
				if (db.Ext().IsCached(pilotID))
				{
					db.Ext().Purge(pilot);
				}
				ObjectSet pilots1 = db.Get(new Pilot(pilotName));
				ListResult(pilots1);
			} 
			finally 
			{
				db.Close();
			}
		}
		// end CheckReferenceCacheWithPurge
	
		public static void TestCopyingWithPurge()
		{
			SetObjects();
			ObjectContainer db = Db4o.OpenFile(YapFileName);
			try 
			{
				ObjectSet pilots = db.Get(typeof(Pilot));
				Pilot pilot = (Pilot)pilots[0];
				db.Ext().Purge(pilot);
				db.Set(pilot);
				pilots = db.Get(typeof(Pilot));
				ListResult(pilots);
			} 
			finally 
			{
				db.Close();
			}
		}
		// end TestCopyingWithPurge
	
		public static void TestBind()
		{
			SetObjects();
			ObjectContainer db = Db4o.OpenFile(YapFileName);
			try 
			{
				Query q = db.Query();
				q.Constrain(typeof(Car));
				q.Descend("_model").Constrain("Ferrari");
				ObjectSet result = q.Execute();
				Car car1 = (Car)result[0];
				long IdCar1 = db.Ext().GetID(car1);
				Car car2 = new Car("BMW", new Pilot("Rubens Barrichello"));
				db.Ext().Bind(car2,IdCar1);
				db.Set(car2);

				result = db.Get(typeof(Car));
				ListResult(result);
			} 
			finally 
			{
				db.Close();
			}
		}
		// end TestBind

		public static void ListResult(ObjectSet result)
		{
			Console.WriteLine(result.Count);
			foreach (object item in result)
			{
				Console.WriteLine(item);
			}
		}
		// end ListResult
	}
}

File: db4on/reference/vb/db4o-reference-chapters/src/com/db4odoc/f1/identity/IdentityExample.vb
' Copyright (C) 2004 - 2006 db4objects Inc. http://www.db4o.com 

Imports System
Imports System.IO
Imports com.db4o
Imports com.db4o.query

Namespace com.db4odoc.f1.identity
    Public Class IdentityExample
        Public Shared ReadOnly YapFileName As String = "formula1.yap"

        Public Shared Sub Main(ByVal args() As String)
            CheckUniqueness()
            CheckReferenceCache()
            CheckReferenceCacheWithPurge()
            TestBind()

            TestCopyingWithPurge()
        End Sub
        ' end Main

        Public Shared Sub SetObjects()
            File.Delete(YapFileName)
            Dim db As ObjectContainer = Db4o.OpenFile(YapFileName)
            Try
                Dim car As Car = New Car("BMW", New Pilot("Rubens Barrichello"))
                db.Set(car)
                car = New Car("Ferrari", New Pilot("Michael Schumacher"))
                db.Set(car)
            Finally
                db.Close()
            End Try
        End Sub
        ' end SetObjects

        Public Shared Sub CheckUniqueness()
            SetObjects()
            Dim db As ObjectContainer = Db4o.OpenFile(YapFileName)
            Try
                Dim cars As ObjectSet = db.Get(GetType(Car))
                Dim car As Car = CType(cars(0), Car)
                Dim pilotName As String = car.Pilot.Name
                Dim pilots As ObjectSet = db.Get(New Pilot(pilotName))
                Dim pilot As Pilot = CType(pilots(0), Pilot)
                System.Console.WriteLine("Retrieved objects are identical: " + (pilot Is car.Pilot).ToString())
            Finally
                db.Close()
            End Try
        End Sub
        ' end CheckUniqueness

        Public Shared Sub CheckReferenceCache()
            SetObjects()
            Dim db As ObjectContainer = Db4o.OpenFile(YapFileName)
            Try
                Dim pilots As ObjectSet = db.Get(GetType(Pilot))
                Dim pilot As Pilot = CType(pilots(0), Pilot)
                Dim pilotName As String = pilot.Name
                pilot.Name = "new name"
                System.Console.WriteLine("Retrieving pilot by name: " + pilotName)
                Dim pilots1 As ObjectSet = db.Get(New Pilot(pilotName))
                ListResult(pilots1)
            Finally
                db.Close()
            End Try
        End Sub
        ' end CheckReferenceCache

        Public Shared Sub CheckReferenceCacheWithPurge()
            SetObjects()
            Dim db As ObjectContainer = Db4o.OpenFile(YapFileName)
            Try
                Dim pilots As ObjectSet = db.Get(GetType(Pilot))
                Dim pilot As Pilot = CType(pilots(0), Pilot)
                Dim pilotName As String = pilot.Name
                pilot.Name = "new name"
                System.Console.WriteLine("Retrieving pilot by name: " + pilotName)
                Dim pilotID As Long = db.Ext().GetID(pilot)
                If db.Ext().IsCached(pilotID) Then
                    db.Ext().Purge(pilot)
                End If
                Dim pilots1 As ObjectSet = db.Get(New Pilot(pilotName))
                ListResult(pilots1)
            Finally
                db.Close()
            End Try
        End Sub
        ' end CheckReferenceCacheWithPurge

        Public Shared Sub TestCopyingWithPurge()
            SetObjects()
            Dim db As ObjectContainer = Db4o.OpenFile(YapFileName)
            Try
                Dim pilots As ObjectSet = db.Get(GetType(Pilot))
                Dim pilot As Pilot = CType(pilots(0), Pilot)
                db.Ext().Purge(pilot)
                db.Set(pilot)
                pilots = db.Get(GetType(Pilot))
                ListResult(pilots)
            Finally
                db.Close()
            End Try
        End Sub
        ' end TestCopyingWithPurge

        Public Shared Sub TestBind()
            SetObjects()
            Dim db As ObjectContainer = Db4o.OpenFile(YapFileName)
            Try
                Dim q As Query = db.Query()
                q.Constrain(GetType(Car))
                q.Descend("_model").Constrain("Ferrari")
                Dim result As ObjectSet = q.Execute()
                Dim car1 As Car = CType(result(0), Car)
                Dim IdCar1 As Long = db.Ext().GetID(car1)
                Dim car2 As Car = New Car("BMW", New Pilot("Rubens Barrichello"))
                db.Ext().Bind(car2, IdCar1)
                db.Set(car2)

                result = db.Get(GetType(Car))
                ListResult(result)
            Finally
                db.Close()
            End Try
        End Sub
        ' end TestBind

        Public Shared Sub ListResult(ByVal result As ObjectSet)
            Console.WriteLine(result.Count)
            For Each item As Object In result
                Console.WriteLine(item)
            Next
        End Sub
        ' end ListResult

    End Class
End Namespace
